File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.py
from collections import defaultdict

n = int(input())
tanghuru = list(map(int, input().split()))
start, end, count = 0, 0, 0
huru_count = defaultdict(int)
answer = 0

while end<n:
    if huru_count[tanghuru[end]]==0:
        count += 1
    huru_count[tanghuru[end]] += 1

    while count>2:
        huru_count[tanghuru[start]]-=1
        if huru_count[tanghuru[start]]==0:
            count -= 1
        start += 1

    answer = max(answer, end-start+1)
    end += 1

print(answer)

File: 백준/Silver/30804. 과일 탕후루/과일 탕후루.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        int start = 0, end = 0;
        HashMap<Integer, Integer> hash = new HashMap<>();
        int answer = 0;

        while (end < n) {
            hash.put(arr[end], hash.getOrDefault(arr[end], 0) + 1);

            while (hash.size() > 2) {
                if (hash.get(arr[start]) == 1) {
                    hash.remove(arr[start]);
                } else {
                    hash.put(arr[start], hash.get(arr[start]) - 1);
                }
                start++;
            }

            answer = Math.max(answer, end - start + 1);
            end++;
        }

        System.out.println(answer);
    }
}